server/src/main/java/org/opensearch/cluster/routing/FailAwareWeightedRouting.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.cluster.routing;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.opensearch.OpenSearchException;
+import org.opensearch.action.search.SearchShardIterator;
+import org.opensearch.cluster.ClusterState;
+import org.opensearch.cluster.metadata.WeightedRoutingMetadata;
+import org.opensearch.cluster.node.DiscoveryNode;
+import org.opensearch.index.shard.ShardId;
+import org.opensearch.rest.RestStatus;
+import org.opensearch.search.SearchShardTarget;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Stream;
+
+/**
+ * This class contains logic to find next shard to retry search request in case of failure from other shard copy.
+ * This decides if retryable shard search requests can be tried on shard copies present in data
+ * nodes whose attribute value weight for weighted shard routing is set to zero.
+ */
+
+public class FailAwareWeightedRouting {
+
+    private static final Logger logger = LogManager.getLogger(FailAwareWeightedRouting.class);
+
+    /* exception causing failure for previous shard copy */
+    private Exception exception;
+
+    private ClusterState clusterState;
+
+    public FailAwareWeightedRouting(Exception e, ClusterState clusterState) {
+        this.exception = e;
+        this.clusterState = clusterState;
+    }
+
+    /**
+     * *
+     * @return true if exception is due to cluster availability issues
+     */
+    private boolean isInternalFailure() {
+        if (exception instanceof OpenSearchException) {
+            // checking for 5xx failures
+            return (((OpenSearchException) exception).status().getStatus() / 100) * 100 == RestStatus.INTERNAL_SERVER_ERROR.getStatus();
```

Comment:
1. Create a Immutable list of exhaustive 5xx codes 
2. Compare the status against items on that list
3. Remove division/mul by 100

Reply:
We can create a list of 5xx codes but that will require regular update. Did this since we are interested in all OpenSearchException with 5xx status code. 

Any specific reason I have missed for which you are suggesting this?